package/MDAnalysis/coordinates/PDB.py

```diff
@@ -1217,10 +1217,30 @@ def validate_chainids(chainids, default):
         else:
             atom_ids = np.arange(len(atoms)) + 1
 
-        for i, atom in enumerate(atoms):
+        def get_atom_name_lookup_dict() -> dict[tuple[str, str], str]:
+            """Creates a lookup dictionary for deduced atom names
+
+            Unique pairs of atomnames and resnames are found first to avoid
+            redundant calls to `_deduce_PDB_atom_name`
+            """
+            pairs = [(atomnames[i], resnames[i])
+                     for i in range(len(atomnames))]
+
+            unique_pairs = set(pairs)
+
+            deduced_PDB_atom_names = {
+                (atom, res): self._deduce_PDB_atom_name(atom, res)
+                for atom, res in unique_pairs
+                }
+
+            return deduced_PDB_atom_names
```

Comment:
Is this not a manual reimplementation of what `@functools.cache` would do for free on `_deduce_PDB_atom_name`? Or maybe I'm missing something

Reply:
That's a great suggestion, thanks very much! I wasn't aware of this decorator.